### DISCUSSION TITLE
Implementing Runno

### DISCUSSION THREAD

**Discussion by LukePrior** [taybenlor/runno#188] [General]
Hi Ben,

I just wanted to thank you for the wonderful and prompt support that you have offered.

### My current prototype

I'm just having a look at your example currently and must say they look very nice, I am going to look into implementing the code marking features that you have also displayed as I think these would be helpful.

You can find my **very** basic prototype here: https://comp1511-guide.vercel.app/ (running on Vercel so that the headers work)

### Styling Ideas

I think I will redo my page to follow the same layout as your example so I have a good base to build from.

The main problem I'm running into currently is the styling issue when code in the editor is longer than the allocated space.

This is how it looks with the default inputs on your site:
![image](https://user-images.githubusercontent.com/22492406/161977451-a017491a-3463-4d05-a7d5-bbe441bc89cd.png)

The problem occurs when too many lines of code are entered:
![image](https://user-images.githubusercontent.com/22492406/161977571-fc0327ac-14fd-441e-8cc7-2c5d9bd6a0a6.png)

The total element size remains the same but the terminal shrinks to accommodate the increased lines of code. I would like the ability to set the total element height and the height of the terminal, the code editor should show a scrollbar if it all cannot be displayed however the current behaviour is this:

![image](https://user-images.githubusercontent.com/22492406/161977831-2fcf8525-e6be-44b6-8311-6d083fe5c69c.png)

I'm sure you agree that this isn't the best way to handle this so I suggest that you update the CSS for `runno-editor` to enable `overflow:auto`. This results in the previous scenario becoming:

![image](https://user-images.githubusercontent.com/22492406/161978285-2d6dc685-0917-4b80-a566-1f248d827d63.png)

This is only marginally better with the element now keeping its specified height however the terminal needs some more formatting so that it is always positioned correctly. I have found that this is most easily done by giving `runno-terminal` a CSS `height` style of above 15px. This results in the following:

![image](https://user-images.githubusercontent.com/22492406/161978945-9b88e7a2-6eb4-4f86-b1a7-e84f93d36f4e.png)

This is now reaching a point which I believe to be very useable but I think the scrollbars for the code editor and terminal could do with some more styling. Not too sure on what to do for this so I will leave that to your discretion.

![image](https://user-images.githubusercontent.com/22492406/161979283-144268ad-a37e-4e95-9de7-03c38396160b.png)

Thanks,
Luke

**Comment by taybenlor** [comment]
Thanks Luke, yeah I'd like to get a solution that works well for a web component, and also works well in the iframe. You can see that with the iframe version the editor collapses and starts scrolling:

<img width="584" alt="Screen Shot 2022-04-07 at 8 01 43 am" src="https://user-images.githubusercontent.com/137128/162079597-6041db6d-2c14-48fa-803f-a6151f4ced2b.png">

The iframe is just a `<runno-run>` element set to a height and width of 100%. So I need to figure out what is different that causes the alternate behaviour outside of the iframe.

**Comment by taybenlor** [comment]
PS: Would love to hear more about your use case. Looks like it's for a programming course?

**Comment by taybenlor** [comment]
I've improved the default styling of the web component so that the terminal has a minimum height, and the editor resizes nicely. This means it's not necessary to set a fixed height on the web component. If you do set a fixed height the two areas will share the space 60:40. There was also a bug I fixed in the styling of the editor that made it take up more than its fair share of space.

The change is in https://github.com/taybenlor/runno/pull/189

I'll cut a release soon as well.

**Comment by taybenlor** [comment]
These changes are available in `@runno/runtime:0.3.1` and documented at https://runno.dev/#component-elements

**Reply by LukePrior** [reply]
Thanks, they were just the thing I was looking for!

**Comment by LukePrior** [comment]
I'm not sure if I'm just doing this wrong or if it's a bug but when I edited your example to use clang instead of Python I'm running into an error with the check code examples:

`const { result } = await provider.headlessRunCode("clang", program, stdin);`

I get the following error when the above code is run:

`runtime.es.js:9 Error: No such file or directory: ./program.wasm`
![image](https://user-images.githubusercontent.com/22492406/162091357-31e46e85-ac39-4ddd-8713-e227db123ac5.png)

**Reply by taybenlor** [reply]
That looks like the code didn't successfully compile (the `program.wasm` is output by clang). What happens if you run the code normally? (not headless)

**Reply by LukePrior** [reply]
The code is just a simple Hello World and runs correctly when run normally.
![image](https://user-images.githubusercontent.com/22492406/162095472-b40bfb2c-6315-4f63-99f2-128babe24201.png)

**Reply by taybenlor** [reply]
Okay that sounds like it's a bug with the way headless running works. I can look at fixing that later (probably won't be tonight).

A workaround would be to run each of the clang steps manually. You'll need to look at [`provider.ts`](https://github.com/taybenlor/runno/blob/main/packages/runtime/src/provider.ts) inside the `runtime` package to see how it works. 

Basically the idea is to run each of the following steps:

```
prepare: [
        `clang -cc1 -Werror -triple wasm32-unkown-wasi -isysroot /sys -internal-isystem /sys/include -ferror-limit 4 -fmessage-length 80 -fcolor-diagnostics -O2 -emit-obj -o ./program.o ${entryPath}`,
        `wasm-ld -L/sys/lib/wasm32-wasi /sys/lib/wasm32-wasi/crt1.o ./program.o -lc -o ./program.wasm`,
],
run: `wasmer run ./program.wasm`,
```

You can do that by using `headlessUnsafeCommand`, collecting the `fs` which is returned and then passing it to the next command.

The bug will be happening because the `fs` is not being passed between steps in this process.

**Reply by LukePrior** [reply]
No rush, just something I stumbled upon during testing.

**Reply by taybenlor** [reply]
Could you file this as an issue? That'll help me keep track of it.

**Comment by LukePrior** [comment]
I've also encountered an issue with detecting EOF (Ctrl D) as the web terminal doesn't seem to register this like a normal terminal would.

You can see the behaviour [here](https://runno.run/?editor=1&runtime=clang&code=I2luY2x1ZGUgPHN0ZGlvLmg-CgojZGVmaW5lIE1BWF9TSVpFIDEwMQoKaW50IG1haW4odm9pZCkgewoKICAgIGludCBudW1iZXJzW01BWF9TSVpFXTsKCiAgICBwcmludGYoIkVudGVyIG51bWJlcnM6IFxuIik7CgogICAgaW50IGkgPSAwOwogICAgd2hpbGUgKHNjYW5mKCIlZCIsICZudW1iZXJzW2ldKSAhPSBFT0YpIHsKICAgICAgICBpKys7CiAgICB9CgogICAgaW50IHkgPSAwOwogICAgd2hpbGUgKHkgPCBpKSB7CiAgICAgICAgcHJpbnRmKCIlZFxuIiwgbnVtYmVyc1t5XSk7CiAgICAgICAgeSsrOwogICAgfQoKICAgIHJldHVybiAwOwp9).

Do you know of any solutions or workarounds, I can't really change the code as it needs to be the same as what runs normally.

**Reply by taybenlor** [reply]
Thanks, yeah okay I think this should be possible. I can't think of a workaround the way things work currently. I'll need to add a handler specifically for `ctrl + D` and end the interactive capture of STDIN.

**Reply by taybenlor** [reply]
Can you file this as an issue?
